Automattic/jetpack
# Blocks: Show Blocks if they're missing a plan, add Upgrade Nudge

Picking up @scruffian's Automattic/wp-calypso#33723 and moving it into JP. 
The rationale is that these blocks live in JP, so it's much easier to handle e.g. minimum plan information coming from the endpoint here. Furthermore, we might eventually want this upgrade nudge to also work in JP (where applicable!).

#### Changes proposed in this Pull Request:

Enable paid blocks on plans that don't include them for insertion into the editor, and add an upgrade nudge to them. Feature-flagged for now.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
See paAmJe-il-p2

#### Screenshot

Styling pending some help from a designer :grimacing: : 

![image](https://user-images.githubusercontent.com/96308/60432556-65ab6d80-9c24-11e9-86f7-eb83ea110ed8.png)

Frontend (pre-existing, not touched by this PR, but will probably see some style and copy polish in a follow-up): 
![image](https://user-images.githubusercontent.com/96308/60431329-5545c380-9c21-11e9-832e-4af52720ce51.png)

#### Testing instructions:

- Switch to this branch locally, build Jetpack (`yarn build`), and start Docker (`yarn docker:up`)
- Make sure you're connected to WP.com, and on a free plan
- Start a new post, and look for the 'Simple Payments' block in the block picker. It shouldn't be there :slightly_smiling_face: 
- Now add the following line to any file in `docker/mu-plugins`, e.g. a newly created `0-blocks-upgrade.php` (needs to start with `<?php`): `define ( 'JETPACK_SHOW_BLOCK_UPGRADE_NUDGE', true );`
- Reload the post editor screen in your browser
- Look for the 'Simple Payments' block in the block picker again, It should be there now: Insert it into the post.
- You should see it slightly faded out, and with an upgrade banner on top of it (see screenshots above)
- Edit some block settings and save the post. When viewing while logged-in, you'll see a banner telling you that your plan doesn't include Simple Payments (see screenshot above). (Note that this message has been there before this PR, in case a user downgraded from a Premium Plan. It isn't touched by this PR but possibly by a follow-up.)
- Back in the editor, click the 'Upgrade' button. Verify that you're taken to the plans page, and the Premium Plan is highlighted as 'Suggested'. Purchase the plan.
- Navigate back to the editor and reload. Verify that the upgrade nudge banner is gone, the block is no longer faded out, and the settings have been persisted from before. Make sure you don't modify the post.
- View the post in the frontend again (you might have to reload). Verify that the block is now present and functional.
- Verify that downgrading to the Free plan restores the original state (with the upgrade nudge).

Once this PR has approval, I'll manually sync the newly created files over to the WP.com counterpart, D29733-code. Doing that now doesn't make sense since they'll be removed by Fusion each time this PR changes.

#### Proposed changelog entry for your changes:

Enable paid blocks on plans that don't include them for insertion into the editor, and add an upgrade nudge to them.

#### TODO

- [x] Hide the block on the frontend if the current plan doesn't include it
  - We're getting this for free: https://github.com/Automattic/jetpack/blob/ee0fcec521f94cb2596ea4a8879350ca098457ae/modules/simple-payments/simple-payments.php#L194-L200. However, we might want to update/polish it a bit.
- [x] ~Use `Card` component~ _(not available in `@wordpress/components`)_
- [x] ~Fix flow to return to editor after purchase~ see below (under the 'Follow-up' heading)
- [x] Get some proper plans strings
  - This is going to be interesting since we can't really use Jetpack's `lib/plans/constants` since that file isn't available on WP.com -- only the contents of the `extensions/` directory are (on the client side, anyway). However, we have access to the REST API, so maybe we can fetch information from there.
  - Fortunately, WP.com has the `/plans` endpoint which should provide the relevant information, so I'm thinking of using `@wordpress/data` and `apiFetch` to obtain the relevant strings (query `product_slug` or `path_slug` field, get `product_name_short` field).
- [x] Feature-flag this

Don't forget that this is WP.com first -- we'll probably even add a check to hide it from JP for now. This is mostly relevant for the upgrade flow.

#### Follow-up

- This will require a Calypso counterpart, to allow us to provide a mechanism that redirect users to the editor (i.e. a given URL) post-purchase. [`checkout-thank-you/index.jsx`](https://github.com/Automattic/wp-calypso/blob/baf195241cb6e5c5c33da6ca2b2138fba7d0a07c/client/my-sites/checkout/checkout-thank-you/index.jsx) is a likely candidate for those changes, see e.g. [`redirectIfThemePurchased`](https://github.com/Automattic/wp-calypso/blob/baf195241cb6e5c5c33da6ca2b2138fba7d0a07c/client/my-sites/checkout/checkout-thank-you/index.jsx#L256-L269). We'll probably have to pass a query arg or cookie.
- Furthermore, it'd be good to polish the [frontend message](https://github.com/Automattic/jetpack/blob/ee0fcec521f94cb2596ea4a8879350ca098457ae/modules/simple-payments/simple-payments.php#L194-L200).
- Fix styling
- Add tracking

/cc @simison @sirreal
